test/unit/builtins/async-await.spec.ts

```diff
@@ -0,0 +1,242 @@
+            return await abc();
+        }
+
+        const awaitingPromise = def();
+        awaitingPromise.then(data => log("resolving awaiting promise", data));
+
+        resolve("resolved data");
+
+        return allLogs;
+
+    `
+        .setTsHeader(promiseTestLib)
+        .expectToEqual(["resolving original promise", "resolved data", "resolving awaiting promise", "resolved data"]);
+});
+
+test("async function returning value is same as non-async function returning promise", () => {
+    util.testFunction`
+        function f(): Promise<number> {
+            return Promise.resolve(42);
+        }
+
+        async function fAsync(): Promise<number> {
+            return 42;
+        }
+
+        const { state: state1, value: value1 } = f() as any;
+        const { state: state2, value: value2 } = fAsync() as any;
+
+        return {
+            state1, value1,
+            state2, value2
+        };
+    `.expectToEqual({
+        state1: 1, // __TS__PromiseState.Fulfilled
+        value1: 42,
+        state2: 1, // __TS__PromiseState.Fulfilled
+        value2: 42,
+    });
+});
+
+test("correctly handles awaited functions rejecting", () => {
+    util.testFunction`
+        const { promise: promise1, reject } = defer<string>();
+        const { promise: promise2 } = defer<string>();
+
+        promise1.then(data => log("resolving promise1", data), reason => log("rejecting promise1", reason));
+        promise2.then(data => log("resolving promise2", data));
+
+        async function abc() {
+            const result1 = await promise1;
+            const result2 = await promise2;
+            return [result1, result2];
+        }
+
+        const awaitingPromise = abc();
+        awaitingPromise.catch(reason => log("awaiting promise was rejected because:", reason));
+
+        reject("test reject");
+
+        return allLogs;
+
+    `
+        .setTsHeader(promiseTestLib)
+        .expectToEqual(["rejecting promise1", "test reject", "awaiting promise was rejected because:", "test reject"]);
+});
+
+test("can call async function at top-level", () => {
+    util.testModule`
+        export let aStarted = false;
+        async function a() {
+            aStarted = true;
+            return 42;
+        }
+
+        a(); // Call async function (but cannot await)
+    `
+        .setOptions({ module: ModuleKind.ESNext, target: ScriptTarget.ES2017 })
+        .expectToEqual({
+            aStarted: true,
+        });
+});
+
+test.each(["await a();", "const b = await a();", "export const b = await a();"])(
+    "cannot await at top-level (%p)",
+    awaitUsage => {
+        util.testModule`
+        async function a() {
+            return 42;
+        }
+
+        ${awaitUsage}
+        export {} // Required to make TS happy, cannot await without import/exports
+    `
+            .setOptions({ module: ModuleKind.ESNext, target: ScriptTarget.ES2017 })
+            .expectToHaveDiagnostics([notAllowedTopLevelAwait.code]);
+    }
+);
```

Comment:
Maybe a test (or a few) testing that await does not have any effect when used on non promises.
e.g.

```
const x = 1;
const y = 2;

test((await x) + y == 3)
```

I am not sure if that is possible, but maybe you can detect awaits like this at compile time and ommit them completely (don't wrap them).

TS emits a warning/info diagnostic for such awaits: https://www.typescriptlang.org/play?#code/MYewdgzgLgBAHjAvDAjAbgFCkrAnkmAJky3GhgC8CBDAd2oEtYEBqGXEgUzgAcQAnWAG8YAX0xA